bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,255 @@
+
+    // The `upstream_peer` method is called when a new upstream connection is required.
+    // This method is responsible for selecting a worker from the load balancer and
+    // creating a new peer with the selected worker.
+    // Here we enqueue the request ID in the worker queue and wait for it to be at the front.
+    // If we are retring the request, we remove the request from the queue first.
+    async fn upstream_peer(
+        &self,
+        session: &mut Session,
+        ctx: &mut Self::CTX,
+    ) -> Result<Box<HttpPeer>> {
+        if ctx.tries > 0 {
+            Self::remove_request_from_queue(Self::get_request_id(session)?).await;
+        }
+
+        // Select the worker in a round-robin fashion
+        let worker = self.0.select(b"", 256).ok_or(Error::new_str("Worker not found"))?;
+
+        // Read request ID from headers
+        let request_id = Self::get_request_id(session)?;
+
+        // Enqueue the request ID in the worker queue
+        // We use a new scope to release the lock after the operation
+        {
+            let mut ctx_guard = QUEUES.write().await;
+            let worker_queue = ctx_guard.entry(worker.clone()).or_insert_with(Vec::new);
+
+            // Limit queue length to MAX_QUEUE_ITEMS requests
+            if worker_queue.len() >= MAX_QUEUE_ITEMS {
+                return Err(Error::new_str("Too many requests in the queue"));
+            }
+
+            worker_queue.push(request_id.to_string());
+        }
+
+        // Wait for the request to be at the front of the queue
+        loop {
+            // We use a new scope for each iteration to release the lock
+            {
+                let ctx_guard = QUEUES.read().await;
+                if let Some(worker_queue) = ctx_guard.get(&worker) {
+                    if worker_queue[0] == request_id {
+                        break;
+                    }
+                } else {
+                    return Err(Error::new_str("Worker not found"));
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+
+        // Set SNI
+        let mut http_peer = HttpPeer::new(worker, false, "".to_string());
+        let peer_opts =
+            http_peer.get_mut_peer_options().ok_or(Error::new(ErrorType::InternalError))?;
+
+        // Timeout settings
+        peer_opts.total_connection_timeout = TIMEOUT_SECS;
+        peer_opts.connection_timeout = CONNECTION_TIMEOUT_SECS;
+        peer_opts.read_timeout = TIMEOUT_SECS;
+        peer_opts.write_timeout = TIMEOUT_SECS;
+        peer_opts.idle_timeout = TIMEOUT_SECS;
+
+        // Enable HTTP/2
+        peer_opts.alpn = ALPN::H2;
+
+        let peer = Box::new(http_peer);
+        Ok(peer)
+    }
+
+    // The `upstream_request_filter` method is called before sending the request to the upstream
+    // server. We use the method to ensure that the correct headers are forwarded for gRPC
+    // requests.
+    async fn upstream_request_filter(
```

Comment:
Similar nit re `///` and the comment content as above. One thing that would be helpful to know is when this is called in relation to other methods (e.g., is it called right after `upstream_peer()`?).